### PR TITLE
gambit: refactor

### PR DESCRIPTION
**Pull request by siraben**
###### Motivation for this change
Refactor and prepare for fixing cross-compilation.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/gambit/build.nix [hunk 1]
```diff
@@ -25,10 +25,11 @@ gccStdenv.mkDerivation rec {
   inherit src version git-version;
   bootstrap = gambit-support.gambit-bootstrap;
 
+  nativeBuildInputs = [ git autoconf ];
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/gambit/build.nix:28]: Do we need git here or do we need to fake the information it exctracts?

Comment (siraben) [reply pkgs/development/compilers/gambit/build.nix:28]: It uses git in the makefile.

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 113405` at 475eafff run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/2b3ed63da94edc37187bd1a2e38b1f48">gambit</a></li></li>
    <li><a href="https://gist.github.com/2f29b350475653de578024838dd3011c">gambit-unstable</a></li></li>
    <li><a href="https://gist.github.com/c7fbfddb1e7554f02ff0a18ab20d93b9">gerbil</a></li></li>
    <li><a href="https://gist.github.com/a3da3b4042a90e58b434d1b7894aa293">gerbil-unstable</a></li></li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/development/compilers/gambit/build.nix:85:3:
```
   |
85 |   buildPhase = ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`configurePhase` should probably contain `runHook preConfigure` and `runHook postConfigure`.

Near pkgs/development/compilers/gambit/build.nix:65:3:
```
   |
65 |   configurePhase = ''
   |   ^
```
</li>
  </ul>
</details>

**Comment by rmcgibbo** [comment]
Looks like the bot OOM'd. I'm working on improving that, so sorry for the noise.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 113405` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
4 packages built:
</summary>
<ul>
<li>gambit</li>
<li>gambit-unstable</li>
<li>gerbil</li>
<li>gerbil-unstable</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
gambit:
<details>

Please consider this feature to be alpha.

A substituteInPlace with an unmatched pattern got detected:
```
substituteStream(): WARNING: pattern '/usr/local/opt/openssl@1.1/lib' doesn't match anything in file 'config.status'
```

Please check the offending substituteInPlace for typos or changes in source.

A likely typo in `git-version` argument was found, did you mean `version`?

Near pkgs/development/compilers/gambit/build.nix:25:10:

```
   |
25 |   inherit src version git-version;
   |          ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/attribute-typo.md
`lgpl21` is a deprecated license, check if project uses `lgpl21Plus` or `lgpl21Only` and change `meta.license` accordingly.

Near pkgs/development/compilers/gambit/gambit-support.nix:33:5:

```
   |
33 |     license     = lib.licenses.lgpl21; # dual, also asl20
   |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
`configurePhase` should probably contain `runHook preConfigure` and `runHook postConfigure`.

Near pkgs/development/compilers/gambit/build.nix:65:3:

```
   |
65 |   configurePhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/development/compilers/gambit/build.nix:85:3:

```
   |
85 |   buildPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
</details>
gambit-unstable:
<details>

Please consider this feature to be alpha.

A substituteInPlace with an unmatched pattern got detected:
```
substituteStream(): WARNING: pattern '/usr/local/opt/openssl@1.1/lib' doesn't match anything in file 'config.status'
```

Please check the offending substituteInPlace for typos or changes in source.

A likely typo in `git-version` argument was found, did you mean `version`?

Near pkgs/development/compilers/gambit/build.nix:25:10:

```
   |
25 |   inherit src version git-version;
   |          ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/attribute-typo.md
`lgpl21` is a deprecated license, check if project uses `lgpl21Plus` or `lgpl21Only` and change `meta.license` accordingly.

Near pkgs/development/compilers/gambit/gambit-support.nix:33:5:

```
   |
33 |     license     = lib.licenses.lgpl21; # dual, also asl20
   |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
`configurePhase` should probably contain `runHook preConfigure` and `runHook postConfigure`.

Near pkgs/development/compilers/gambit/build.nix:65:3:

```
   |
65 |   configurePhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/development/compilers/gambit/build.nix:85:3:

```
   |
85 |   buildPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
</details>
gerbil:
<details>


`lgpl21` is a deprecated license, check if project uses `lgpl21Plus` or `lgpl21Only` and change `meta.license` accordingly.

Near pkgs/development/compilers/gerbil/build.nix:95:5:

```
   |
95 |     license     = lib.licenses.lgpl21; # also asl20, like Gambit
   |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
`configurePhase` should probably contain `runHook preConfigure` and `runHook postConfigure`.

Near pkgs/development/compilers/gerbil/build.nix:50:3:

```
   |
50 |   configurePhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
Unused argument: `bash`.
Near pkgs/development/compilers/gerbil/build.nix:1:36:

```
  |
1 | { pkgs, gccStdenv, lib, coreutils, bash, # makeStaticLibraries,
  |                                    ^
```
</details>
gerbil-unstable:
<details>


`lgpl21` is a deprecated license, check if project uses `lgpl21Plus` or `lgpl21Only` and change `meta.license` accordingly.

Near pkgs/development/compilers/gerbil/build.nix:95:5:

```
   |
95 |     license     = lib.licenses.lgpl21; # also asl20, like Gambit
   |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
`configurePhase` should probably contain `runHook preConfigure` and `runHook postConfigure`.

Near pkgs/development/compilers/gerbil/build.nix:50:3:

```
   |
50 |   configurePhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
Unused argument: `bash`.
Near pkgs/development/compilers/gerbil/build.nix:1:36:

```
  |
1 | { pkgs, gccStdenv, lib, coreutils, bash, # makeStaticLibraries,
  |                                    ^
```
</details>
</details>

File: pkgs/development/compilers/gambit/build.nix [hunk 2]
```diff
@@ -25,10 +25,11 @@ gccStdenv.mkDerivation rec {
   inherit src version git-version;
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/gambit/build.nix:25]:
```suggestion
  inherit version git-version;
```

File: pkgs/development/compilers/gambit/build.nix [hunk 3]
```diff
@@ -83,15 +84,15 @@ gccStdenv.mkDerivation rec {
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/gambit/build.nix:84]: /usr/local/opt/openssl@1.1/lib 3 lines above here does not match anything and should be removed or fixed.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by 7c6f434c** [comment]
Also @fare I guess

File: pkgs/development/compilers/gambit/build.nix (hunk 2 above)
Comment (siraben) [reply pkgs/development/compilers/gambit/build.nix:25]: It is necessary to inherit src.

Comment (SuperSandro2000) [reply pkgs/development/compilers/gambit/build.nix:25]: Then this is a bug in hammering.

**Comment by siraben** [comment]
Non-blocking, but asked downstream to release a new version https://github.com/gambit/gambit/issues/667 since it's been over 2 years.

File: pkgs/development/compilers/gambit/build.nix (hunk 3 above)
Comment (siraben) [reply pkgs/development/compilers/gambit/build.nix:84]: I think they match https://github.com/gambit/gambit/blob/6acd87dfa95bfca33082a431e72f023345dc07ee/configure.ac#L2041?

Comment (siraben) [reply pkgs/development/compilers/gambit/build.nix:83]: Ok the second substitution doesn't match anything but the first one does.

Comment (siraben) [reply pkgs/development/compilers/gambit/build.nix:84]: On inspecting the build logs, it was using openssl in `/usr/local/lib`. I have fixed the substitutions to make sure this doesn't happen.

File: pkgs/development/compilers/gambit/build.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/compilers/gambit/build.nix:28]: Important is how. If it uses it to extract version information that will not work most likely.

Comment (siraben) [reply pkgs/development/compilers/gambit/build.nix:28]: Please see the makefile https://github.com/gambit/gambit/blob/master/makefile.in

Comment (siraben) [reply pkgs/development/compilers/gambit/build.nix:28]: It creates archives and so on with git.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 113405` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
4 packages built:
</summary>
<ul>
<li>gambit</li>
<li>gambit-unstable</li>
<li>gerbil</li>
<li>gerbil-unstable</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
gambit-unstable:
<details>

Please consider this feature to be alpha.

A substituteInPlace with an unmatched pattern got detected:
```
substituteStream(): WARNING: pattern '/usr/local/opt/openssl@1.1' doesn't match anything in file 'config.status'
```

Please check the offending substituteInPlace for typos or changes in source.
</details>
gambit:
<details>

Please consider this feature to be alpha.

A substituteInPlace with an unmatched pattern got detected:
```
substituteStream(): WARNING: pattern '/usr/local/opt/openssl@1.1' doesn't match anything in file 'config.status'
```

Please check the offending substituteInPlace for typos or changes in source.
</details>
gerbil-unstable:
<details>


Unused argument: `bash`.
Near pkgs/development/compilers/gerbil/build.nix:1:36:

```
  |
1 | { pkgs, gccStdenv, lib, coreutils, bash, # makeStaticLibraries,
  |                                    ^
```
</details>
gerbil:
<details>


Unused argument: `bash`.
Near pkgs/development/compilers/gerbil/build.nix:1:36:

```
  |
1 | { pkgs, gccStdenv, lib, coreutils, bash, # makeStaticLibraries,
  |                                    ^
```
</details>
</details>

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 113405` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
4 packages built:
</summary>
<ul>
<li>gambit</li>
<li>gambit-unstable</li>
<li>gerbil</li>
<li>gerbil-unstable</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
gambit:
<details>

Please consider this feature to be alpha.

A substituteInPlace with an unmatched pattern got detected:
```
substituteStream(): WARNING: pattern '/usr/local/opt/openssl@1.1' doesn't match anything in file 'config.status'
```

Please check the offending substituteInPlace for typos or changes in source.

A likely typo in `git-version` argument was found, did you mean `version`?

Near pkgs/development/compilers/gambit/build.nix:25:10:

```
   |
25 |   inherit src version git-version;
   |          ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/attribute-typo.md
`lgpl21` is a deprecated license, check if project uses `lgpl21Plus` or `lgpl21Only` and change `meta.license` accordingly.

Near pkgs/development/compilers/gambit/gambit-support.nix:33:5:

```
   |
33 |     license     = lib.licenses.lgpl21; # dual, also asl20
   |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
`configurePhase` should probably contain `runHook preConfigure` and `runHook postConfigure`.

Near pkgs/development/compilers/gambit/build.nix:65:3:

```
   |
65 |   configurePhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/development/compilers/gambit/build.nix:85:3:

```
   |
85 |   buildPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
</details>
gambit-unstable:
<details>

Please consider this feature to be alpha.

A substituteInPlace with an unmatched pattern got detected:
```
substituteStream(): WARNING: pattern '/usr/local/opt/openssl@1.1' doesn't match anything in file 'config.status'
```

Please check the offending substituteInPlace for typos or changes in source.

A likely typo in `git-version` argument was found, did you mean `version`?

Near pkgs/development/compilers/gambit/build.nix:25:10:

```
   |
25 |   inherit src version git-version;
   |          ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/attribute-typo.md
`lgpl21` is a deprecated license, check if project uses `lgpl21Plus` or `lgpl21Only` and change `meta.license` accordingly.

Near pkgs/development/compilers/gambit/gambit-support.nix:33:5:

```
   |
33 |     license     = lib.licenses.lgpl21; # dual, also asl20
   |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
`configurePhase` should probably contain `runHook preConfigure` and `runHook postConfigure`.

Near pkgs/development/compilers/gambit/build.nix:65:3:

```
   |
65 |   configurePhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/development/compilers/gambit/build.nix:85:3:

```
   |
85 |   buildPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
</details>
gerbil:
<details>


`lgpl21` is a deprecated license, check if project uses `lgpl21Plus` or `lgpl21Only` and change `meta.license` accordingly.

Near pkgs/development/compilers/gerbil/build.nix:95:5:

```
   |
95 |     license     = lib.licenses.lgpl21; # also asl20, like Gambit
   |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
`configurePhase` should probably contain `runHook preConfigure` and `runHook postConfigure`.

Near pkgs/development/compilers/gerbil/build.nix:50:3:

```
   |
50 |   configurePhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
Unused argument: `bash`.
Near pkgs/development/compilers/gerbil/build.nix:1:36:

```
  |
1 | { pkgs, gccStdenv, lib, coreutils, bash, # makeStaticLibraries,
  |                                    ^
```
</details>
gerbil-unstable:
<details>


`lgpl21` is a deprecated license, check if project uses `lgpl21Plus` or `lgpl21Only` and change `meta.license` accordingly.

Near pkgs/development/compilers/gerbil/build.nix:95:5:

```
   |
95 |     license     = lib.licenses.lgpl21; # also asl20, like Gambit
   |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
`configurePhase` should probably contain `runHook preConfigure` and `runHook postConfigure`.

Near pkgs/development/compilers/gerbil/build.nix:50:3:

```
   |
50 |   configurePhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
Unused argument: `bash`.
Near pkgs/development/compilers/gerbil/build.nix:1:36:

```
  |
1 | { pkgs, gccStdenv, lib, coreutils, bash, # makeStaticLibraries,
  |                                    ^
```
</details>
</details>

**Comment by siraben** [comment]
Good to merge?

**Comment by fare** [comment]
LGTM

**Comment by SuperSandro2000** [comment]
> A substituteInPlace with an unmatched pattern got detected:
> substituteStream(): WARNING: pattern '/usr/local/opt/openssl@1.1/lib' doesn't match anything in file 'config.status'

I think we should take a look at this if this is a regression or can be dropped.

**Comment by siraben** [comment]
> I think we should take a look at this if this is a regression or can be dropped.

I couldn't find the string in the config file, so I removed it.

**Comment by fare** [comment]
Grepping around there are still these that need to be addressed:
```
configure:             OPENSSL_DIR="/usr/local/opt/openssl@1.1"
configure:             LIBS="$LIBS -L/usr/local/opt/openssl/lib -lssl -lcrypto"
configure.ac:             OPENSSL_DIR="/usr/local/opt/openssl@1.1"
configure.ac:             LIBS="$LIBS -L/usr/local/opt/openssl/lib -lssl -lcrypto"
```

**Comment by fare** [comment]
Also mind that there are two packages, `gambit` and `gambit-unstable`.

**Comment by fare** [comment]
Finally, I have plenty of updates in #114449 that may or may not conflict with these.

**Comment by siraben** [comment]
> Finally, I have plenty of updates in #114449 that may or may not conflict with these.

Ok, that looks pretty involved, it may be easier to have this PR merged then that one rebased on top.

**Comment by fare** [comment]
OK, I merged your PR into my own #114449.

**Comment by fare** [comment]
FTR, this PR introduced a big regression by dropping the full path from the CC variables saved in Gambit, causing non-hermetic calls from Gambit to the wrong CC from $PATH, when it should have been using the same CC as it was compiled with. This works:
```
  export CC=${gccStdenv.cc.targetPrefix}/bin/gcc
```
Of course, it won't work in case of cross-compilation, since in that case, the `CC` on the target platform will *not* be the same as the `CC` to compile Gambit itself. I have no idea how nixpkgs deals with toolchain bootstrap, but special magic will be required here, too.

**Comment by fare** [comment]
And yes, I will fix the regression in my ongoing PR, but there is no easy solution to cross-compiled bootstrap. Executing binaries of both architectures (possibly via qemu) will be required.

**Pull request merged**